File: 377.组合总和-ⅳ.cpp
/*
 * @lc app=leetcode.cn id=377 lang=cpp
 *
 * [377] 组合总和 Ⅳ
 */
#include <vector>
#include <iostream>
#include <numeric>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<int> path;
    int result;
    int combinationSum4(vector<int>& nums, int target) {
        for (int i = 0; i < nums.size(); i++){
            path.push_back(nums[i]);
            backtracking(nums, target);
            path.pop_back();
        }
        return result;
    }

    void backtracking(vector<int>& nums, int target){
        int current_sum = accumulate(path.begin(), path.end(), 0);
        if (current_sum == target){
            // for (auto x:path){
            //     cout << x << " ";
            // }
            // cout << endl;
            result ++;
            return;
        }
        if (current_sum > target){
            return;
        }
        for (int i = 0; i < nums.size(); i++){
            path.push_back(nums.at(i));
            backtracking(nums, target);
            path.pop_back();
        }
    }

    
};
// @lc code=end

File: first_time/DP/zeroonepack/377.组合总和-ⅳ.py
#
# @lc app=leetcode.cn id=377 lang=python3
#
# [377] 组合总和 Ⅳ
#
from typing import List
# @lc code=start
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [0 for _ in range(target+1)]
        dp[0] = 1
        for i in range(0, target+1):
            for j in range(len(nums)):
                if i >= nums[j]:
                    dp[i] += dp[i-nums[j]]
        return dp[-1]
# @lc code=end

